src/lib/keccak.ts

```diff
@@ -64,14 +102,156 @@ const ROUND_CONSTANTS = [
+      const wordBytes = bytestringArray.slice(idx, idx + BYTES_PER_WORD);
+
+      for (let z = 0; z < BYTES_PER_WORD; z++) {
+        // Field element containing value 2^(8*z)
+        const shift = Field.from(2n ** BigInt(8 * z));
+        state[x][y] = state[x][y].add(shift.mul(wordBytes[z]));
+      }
+    }
+  }
+  return state;
+}
+
+// Converts a state of Fields to a list of bytes as Fields and creates constraints for it
+function keccakStateToBytes(state: Field[][]): Field[] {
+  const stateLengthInBytes = KECCAK_STATE_LENGTH / 8;
+  const bytestring: Field[] = Array.from(
+    { length: stateLengthInBytes },
+    (_, idx) =>
+      existsOne(() => {
+        // idx = z + 8 * ((dim * y) + x)
+        const z = idx % BYTES_PER_WORD;
+        const x = Math.floor(idx / BYTES_PER_WORD) % KECCAK_DIM;
+        const y = Math.floor(idx / BYTES_PER_WORD / KECCAK_DIM);
+        // [7 6 5 4 3 2 1 0] [x=0,y=1] [x=0,y=2] [x=0,y=3] [x=0,y=4]
+        //        [x=1,y=0] [x=1,y=1] [x=1,y=2] [x=1,y=3] [x=1,y=4]
+        //        [x=2,y=0] [x=2,y=1] [x=2,y=2] [x=2,y=3] [x=2,y=4]
+        //        [x=3,y=0] [x=3,y=1] [x=3,y=2] [x=3,y=3] [x=3,y=4]
+        //        [x=4,y=0] [x=4,y=1] [x=4,y=0] [x=4,y=3] [x=4,y=4]
+        const word = state[x][y].toBigInt();
+        const byte = (word >> BigInt(8 * z)) & BigInt('0xff');
+        return byte;
+      })
+  );
+
+  // Check all words are composed correctly from bytes
+  for (let y = 0; y < KECCAK_DIM; y++) {
+    for (let x = 0; x < KECCAK_DIM; x++) {
+      const idx = BYTES_PER_WORD * (KECCAK_DIM * y + x);
+      // Create an array containing the 8 bytes starting on idx that correspond to the word in [x,y]
+      const word_bytes = bytestring.slice(idx, idx + BYTES_PER_WORD);
+      // Assert correct decomposition of bytes from state
+      checkBytesToWord(state[x][y], word_bytes);
+    }
+  }
+  return bytestring;
+}
+
+// XOR two states together and return the result
+function keccakStateXor(a: Field[][], b: Field[][]): Field[][] {
+  assert(
+    a.length === KECCAK_DIM && a[0].length === KECCAK_DIM,
+    'Invalid a dimensions'
+  );
+  assert(
+    b.length === KECCAK_DIM && b[0].length === KECCAK_DIM,
+    'Invalid b dimensions'
+  );
+
+  // Calls Gadgets.xor on each pair (x,y) of the states input1 and input2 and outputs the output Fields as a new matrix
+  return a.map((row, rowIndex) =>
+    row.map((element, columnIndex) =>
+      Gadgets.xor(element, b[rowIndex][columnIndex], 64)
+    )
+  );
+}
+
 // KECCAK HASH FUNCTION
 
+// Computes the number of required extra bytes to pad a message of length bytes
+function bytesToPad(rate: number, length: number): number {
+  return Math.floor(rate / 8) - (length % Math.floor(rate / 8));
+}
+
+// Pads a message M as:
+// M || pad[x](|M|)
+// Padding rule 0x06 ..0*..1.
+// The padded message vector will start with the message vector
+// followed by the 0*1 rule to fulfill a length that is a multiple of rate (in bytes)
```

Comment:
Maybe remove those lines

// The padded message vector will start with the message vector
// followed by the 0*1 rule to fulfill a length that is a multiple of rate (in bytes)

and just include the following

// (This means a 0110 sequence, followed with as many 0s as needed, and a final 1 bit)